spec/abci++/v2.md

```diff
@@ -0,0 +1,180 @@
+# Tendermint v2 Markdown pseudocode
+
+This adds a single-threaded implementation of ABCI++,
+with no optimization for splitting out verifying the header and verifying the proposal.
+
+### Initialization
+
+```go
+h_p ← 0
+round_p ← 0
+step_p is one of {propose, prevote, precommit}
+decision_p ← Vector()
+lockedValue_p ← nil
+validValue_p ← nil
+validRound_p ← -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        } else {
+            txdata ← mempool.GetBlock()
+            // getUnpreparedBlockProposal takes tx data, and fills in the unprepared header data
+            unpreparedProposal ← getUnpreparedBlockProposal(txdata)
+            // ABCI++: the proposer may reorder/update transactions in `unpreparedProposal`
+            proposal ← ABCI.PrepareProposal(unpreparedProposal)
```

Comment:
Might be good idea adding note that `PrepareProposal` is adding latency on the critical path, and therefore, `timeoutPropose` would need to be updated accordingly. One idea would be allowing application to provide some hint to the Tendermint regarding duration of `PrepareProposal` so the `timeoutPropose` could be estimated more correctly.